libbeat/processors/dissect/dissect_test.go

```diff
@@ -0,0 +1,282 @@
+			"server":   "machine-321",
+		},
+	},
+	{
+		Name: "when the delimiters contains `{` and `}`",
+		Tok:  "{%{a}}{%{b}} %{rest}",
+		Msg:  "{c}{d} anything",
+		Expected: Map{
+			"a":    "c",
+			"b":    "d",
+			"rest": "anything",
+		},
+	},
+}
+
+func TestDissect(t *testing.T) {
+	if export != nil && *export {
+		dumpJSON()
+		return
+	}
+
+	for _, test := range tests {
+		if test.Skip {
+			continue
+		}
+		t.Run(test.Name, func(t *testing.T) {
+			d, err := New(test.Tok)
+			if !assert.NoError(t, err) {
+				return
+			}
+
+			r, err := d.Dissect(test.Msg)
+			if !assert.NoError(t, err) {
+				return
+			}
+
+			assert.Equal(t, test.Expected, r)
+		})
+	}
+}
+
+var results Map
+var o [][]string
+
+func BenchmarkDissect(b *testing.B) {
+	for _, test := range tests {
+		if test.Skip {
+			continue
+		}
+		b.Run(test.Name, func(b *testing.B) {
+			tok := test.Tok
+			msg := test.Msg
+			d, err := New(tok)
+			if !assert.NoError(b, err) {
+				return
+			}
+			b.ReportAllocs()
+			for n := 0; n < b.N; n++ {
+
+				r, err := d.Dissect(msg)
+				if !assert.NoError(b, err) {
+					return
+				}
+				results = r
+			}
+		})
+	}
+
+	// Add a few regular expression matches agains the same string the test suite,
+	// this give us a baseline to compare to, note that we only test a raw match against the string.
+	b.Run("Regular expression", func(b *testing.B) {
+		re := regexp.MustCompile("/var/log/([a-z]+)/log/([a-z]+)/apache/([a-b]+)")
+		by := "/var/log/docker/more/apache/super"
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			o = re.FindAllStringSubmatch(by, -1)
+		}
+	})
+
+	b.Run("Larger regular expression", func(b *testing.B) {
+		re := regexp.MustCompile("^(\\d{2})-(\\w{3})-(\\d{4})\\s([0-9:.]+)\\s(\\w+)\\s\\[([a-zA-Z0-9-]+)\\]\\s([a-zA-Z0-9.]+)\\s(.+)")
+
+		by := `18-Apr-2018 06:53:20.411 INFO [http-nio-8080-exec-1] org.apache.coyote.http11.Http11Processor.service Error parsing HTTP request header
+ Note: further occurrences of HTTP header parsing errors will be logged at DEBUG level.
+ java.lang.IllegalArgumentException: Invalid character found in method name. HTTP method names must be tokens
+    at org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:426)
+    at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:687)
+    at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
+    at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:790)
+    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1459)
+    at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
+    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
+    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
+    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
+    at java.lang.Thread.run(Thread.java:748)`
```

Comment:
@ph i would be interested to know what would be the performance impact of extracting something from the very end of this long string.

Reply:
Since this code has to read all the bytes to find delimiters and keys the performance will be dependant on the following things:

- Length of the string;
- Numbers of keys;
- appending vs. simple key extraction.

If we take a string as long as the current stack trace to extract one key at the end, assuming we have some repetitive patterns allowing us to create a tokenizer we will have theses results.

With dissect: `"MACHINE[%{key}]"` => 350 ns/op  
Only a regexp match: `regexp.MustCompile("MACHINE\\[(\\w+)\\]$")` =>  656 ns/op  

It will still faster, in that case, assuming we have an easy pattern.

**Note:** in the regexp I am giving a hint that this pattern is a the end of the line, we cannot do that with `dissect` right now with the current syntax.